hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -0,0 +1,506 @@
+          && Objects.equals(sparkMaster, config.sparkMaster)
+          && Objects.equals(sparkMemory, config.sparkMemory)
+          && Objects.equals(assumeDatePartitioning, config.assumeDatePartitioning)
+          && Objects.equals(propsFilePath, config.propsFilePath)
+          && Objects.equals(configs, config.configs);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(basePath, continuous, validateLatestFileSlices, validateLatestBaseFiles, validateAllFileGroups,
+          minValidateIntervalSeconds, sparkMaster, sparkMemory, assumeDatePartitioning, propsFilePath, configs, help);
+    }
+  }
+
+  public static void main(String[] args) {
+    final Config cfg = new Config();
+    JCommander cmd = new JCommander(cfg, null, args);
+
+    if (cfg.help || args.length == 0) {
+      cmd.usage();
+      System.exit(1);
+    }
+
+    SparkConf sparkConf = UtilHelpers.buildSparkConf("Hoodie-Metadata-Table-Validator", cfg.sparkMaster);
+    sparkConf.set("spark.executor.memory", cfg.sparkMemory);
+    JavaSparkContext jsc = new JavaSparkContext(sparkConf);
+
+    HoodieMetadataTableValidator validator = new HoodieMetadataTableValidator(jsc, cfg);
+
+    try {
+      validator.run();
+    } catch (Throwable throwable) {
+      LOG.error("Fail to do hoodie metadata table validation for " + validator.cfg, throwable);
+    } finally {
+      jsc.stop();
+    }
+  }
+
+  public void run() {
+    try {
+      LOG.info(cfg);
+      if (cfg.continuous) {
+        LOG.info(" ****** do hoodie metadata table validation in CONTINUOUS mode ******");
+        doHoodieMetadataTableValidationContinuous();
+      } else {
+        LOG.info(" ****** do hoodie metadata table validation once ******");
+        doHoodieMetadataTableValidationOnce();
+      }
+    } catch (Exception e) {
+      throw new HoodieException("Unable to do hoodie metadata table validation in " + cfg.basePath, e);
+    } finally {
+
+      if (asyncMetadataTableValidateService.isPresent()) {
+        asyncMetadataTableValidateService.get().shutdown(true);
+      }
+    }
+  }
+
+  private void doHoodieMetadataTableValidationOnce() {
+    doMetaTableValidation();
+  }
+
+  private void doHoodieMetadataTableValidationContinuous() {
+    asyncMetadataTableValidateService.ifPresent(service -> {
+      service.start(null);
+      try {
+        service.waitForShutdown();
+      } catch (Exception e) {
+        throw new HoodieException(e.getMessage(), e);
+      }
+    });
+  }
+
+  public void doMetaTableValidation() {
+    metaClient.reloadActiveTimeline();
+    String basePath = metaClient.getBasePath();
+    HoodieSparkEngineContext engineContext = new HoodieSparkEngineContext(jsc);
+
+    List<String> allPartitions = validatePartitions(engineContext, basePath);
+
+    for (String partitionPath : allPartitions) {
+      validateFilesInPartition(engineContext, partitionPath);
+    }
```

Comment:
Can this be parallelized in Spark?  Also, for each partition, `HoodieTableFileSystemView` instances are re-initialized, which can be reused.

Reply:
Sure thing. Changed. Thanks a lot for your review!